flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,305 @@
+import org.apache.flink.ml.metrics.distances.DistanceMetric
+
+import scala.collection.mutable.ListBuffer
+import scala.collection.mutable.PriorityQueue
+
+/**
+ * n-dimensional QuadTree data structure; partitions
+ * spatial data for faster queries (e.g. KNN query)
+ * The skeleton of the data structure was initially
+ * based off of the 2D Quadtree found here:
+ * http://www.cs.trinity.edu/~mlewis/CSCI1321-F11/Code/src/util/Quadtree.scala
+ *
+ * Many additional methods were added to the class both for
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec
+ * @param maxVec
+ */
+class QuadTree(minVec:ListBuffer[Double], maxVec:ListBuffer[Double],distMetric:DistanceMetric){
+  var maxPerBox = 20
+
+  class Node(c:ListBuffer[Double],L:ListBuffer[Double], var children:ListBuffer[Node]) {
+
+    var objects = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return
+      */
+    def getCenterLength(): (ListBuffer[Double], ListBuffer[Double]) = {
+      (c, L)
+    }
+
+    def contains(obj: Vector): Boolean = {
+      overlap(obj, 0.0)
+    }
+
+    /** Tests if obj is within a radius of the node
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def overlap(obj: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) - radius < c(i) + L(i) / 2 && obj(i) + radius > c(i) - L(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == obj.size) {
+        return true
+      } else {
+        return false
+      }
+    }
+
+    /** Tests if obj is near a node:  minDist is defined so that every point in the box
+      * has distance to obj greater than minDist
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < c(i) - L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) + L(i) / 2, 2)
+        } else if (obj(i) > c(i) + L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) - L(i) / 2, 2)
+        }
+      }
+      return minDist
+    }
+
+    def whichChild(obj:Vector):Int = {
+
+      var count = 0
+      for (i <- 0 to obj.size - 1){
+        if (obj(i) > c(i)) {
+          count += Math.pow(2,i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      var cBuff = new ListBuffer[ListBuffer[Double]]
+      cBuff += c
+      var Childrennodes = new ListBuffer[Node]
+      val cPart = partitionBox(cBuff,L,L.length)
+      for (i <- cPart.indices){
+        Childrennodes = Childrennodes :+ new Node(cPart(i), L.map(x => x/2.0), null)
+
+      }
```

Comment:
@chiwanpark @tillrohrmann , sorry for the delay, but I had a look again at making the changes Till suggested.  In the case of multiplying a vector by a double for the line of code ` +        Childrennodes = Childrennodes :+ new Node(cPart(i), L.map(x => x/2.0), null) ` Chiwan posted a workable solution.  

On line 167 of QuadTree, there is also the statement:
``` scala 
val root = new Node( (minVec, maxVec).zipped.map(_ + _).map(x=>0.5*x),
    (maxVec, minVec).zipped.map(_ - _),null) 
```
In pseudo-code, all that is needed is `(0.5*(MinVec + MaxVec), MaxVec - MinVec)`  Addition and multiplication by a double are not permitted for Vectors at this moment.  I could try to implement what Chiwan did for multiplying a double, but that can get ugly and a lot of lines of code.  I'm not sure if we should try to add these operations into the Vector class with this PR, or apply Chiwan's idea to this case as well.